src/haz3lcore/statics/Statics.re

```diff
@@ -459,24 +570,33 @@ and uexp_to_info_map =
       );
     let pat_ms = List.map(((_, _, m)) => m, pat_infos);
     let branch_ms = List.map(((_, _, m)) => m, branch_infos);
-    let branch_frees = List.map(((_, free, _)) => free, branch_infos);
+    let branch_frees =
+      List.map2(
+        ((_, body_free, _), (_, ctx_pat, _)) => {
+          Ctx.subtract_typ(ctx_pat, body_free)
+        },
+        branch_infos,
+        pat_infos,
+      );
     let self = Typ.Joined(Fun.id, branch_sources);
-    let free = Ctx.union([free_scrut] @ branch_frees);
+    /* let free = Ctx.union([free_scrut] @ branch_frees); */
+    let free = Ctx.union([free_scrut, ...branch_frees]);
     add(~self, ~free, union_m([m_scrut] @ pat_ms @ branch_ms));
   };
 }
 and upat_to_info_map =
     (
       ~ctx=Ctx.empty,
       ~mode: Typ.mode=Typ.Syn,
+      ~body_ids: list(Id.t)=[Id.invalid],
```

Comment:
these are the uses, correct? if so, i think they should just be called that, body_ids is vague. also why does it contain Id.invalid by default instead of being empty?